stdlib/public/SDK/XPC/XPCDecoder.swift

```diff
@@ -0,0 +1,110 @@
+
+    public init(withUnderlyingMessage message: xpc_object_t, at codingPath: [CodingKey] = []) {
+        self.underlyingMessage = message
+        self.codingPath = codingPath
+    }
+
+    public func container<Key>(keyedBy type: Key.Type) throws -> KeyedDecodingContainer<Key> where Key : CodingKey {
+        let container = try XPCKeyedDecodingContainer<Key>(referencing: self, wrapping: self.underlyingMessage)
+        return KeyedDecodingContainer(container)
+    }
+
+    public func unkeyedContainer() throws -> UnkeyedDecodingContainer {
+        return try XPCUnkeyedDecodingContainer(referencing: self, wrapping: self.underlyingMessage)
+    }
+
+    public func singleValueContainer() throws -> SingleValueDecodingContainer {
+        return XPCSingleValueDecodingContainer(referencing: self, wrapping: self.underlyingMessage)
+    }
+
+    public static func decode<T: Decodable>(_ type: T.Type, message xpcObject: xpc_object_t) throws -> T {
+        return try T(from: XPCDecoder(withUnderlyingMessage: xpcObject))
+    }
+}
+
+struct XPCDecodingHelpers {
+    static func decodeNil(from xpcObject: xpc_object_t, at codingPath: [CodingKey]) -> Bool {
+        let nullSingleton = xpc_null_create()
+
+        return xpcObject === nullSingleton
+    }
+
+    static func decodeBool(from xpcObject: xpc_object_t, at codingPath: [CodingKey]) throws -> Bool {
+        guard xpc_get_type(xpcObject) == XPC_TYPE_BOOL else {
+            throw DecodingError.typeMismatch(Bool.self,
+                                             DecodingError.Context(codingPath: codingPath,
+                                                                   debugDescription: "Type mismatch.",
+                                                                   underlyingError: nil))
+        }
+
+        return xpcObject === XPC_BOOL_TRUE
+    }
+
+    static func decodeSignedInteger<I: SignedInteger>(_ to: I.Type, from xpcObject: xpc_object_t, at codingPath: [CodingKey]) throws -> I {
+        guard xpc_get_type(xpcObject) == XPC_TYPE_INT64 else {
+            throw DecodingError.typeMismatch(to.self,
+                                             DecodingError.Context(codingPath: codingPath,
+                                                                   debugDescription: "Type mismatch.",
+                                                                   underlyingError: nil))
+        }
+
+        return to.init(exactly: xpc_int64_get_value(xpcObject))!
+    }
+
+    static func decodeUnsignedInteger<U: UnsignedInteger>(_ to: U.Type, from xpcObject: xpc_object_t, at codingPath: [CodingKey]) throws -> U {
+        guard xpc_get_type(xpcObject) == XPC_TYPE_UINT64 else {
+            throw DecodingError.typeMismatch(to.self,
+                                             DecodingError.Context(codingPath: codingPath,
+                                                                   debugDescription: "Type mismatch.",
```

Comment:
It would require to add a dictionary that maps the types to their string representation. I am not sure what the policy on adding runtime data for debugging purposes is, so I am gonna wait on this one.